### DISCUSSION TITLE
Async?

### DISCUSSION THREAD

**Discussion by overtongeist** [FractalFir/rustc_codegen_clr#31] [General]
Has this been tested with async stuff? So, like, any tokio-based logic?

**Comment by FractalFir** [comment]
Not yet. 
Since async requires a runtime (even a very simple one) testing/developing it has been a bit difficult. There is also no multithreading support yet. 

Currently, I am focused on squashing most of the bugs in sync code. 

Besides that, tokio is unlikely to work unmodified, since async stuff (like IO) works differently in .NET.

**Reply by overtongeist** [reply]
Thanks for the reply! When do you think, roughly, that async work would be tackled?

Congrats for the project btw, this is the coolest repo I've found in years. What a time to be alive!

**Reply by FractalFir** [reply]
I have looked into it, and it seems like it mostly boils down to adding support for the [Coroutine](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/ty_kind/enum.TyKind.html#variant.Coroutine) type. This is not trivial, but still relatively easy, and could be done in a couple of days at most. It is very similar to a closure, which I implemented in two days.

The problem is mostly testing - I don't like to have any untested features. I don't want to say async is working, only for that to turn out to not be true.

Exact dates also depend on a lot of IRL stuff (school and final exams), so I can give anything concrete, but I would expect async to be supported before the end of 2024 at worst. However, it could arrive much quicker if all goes to plan.

I am also toying with the idea of posting some development videos to YouTube. I would like to try doing something like what [Tscoding](https://www.youtube.com/@TsodingDaily) or [Andreas Kling](https://youtube.com/playlist?list=PLMOpZvQB55bcTsXzRcBjHJJb52OGWwNjr&feature=shared) do. If this works out for me, async will be the second thing I make videos on, after implementing some debugging tools. In such case, async support could arrive as early as within a month.

So, in short - hard to say, but most likely early this year.
